pyteal/ast/router.py

```diff
@@ -259,13 +263,203 @@ class CondNode:
+    """
+
+    method_sig: str
+    condition: Expr | int
+    method: ABIReturnSubroutine
+
+    def to_cond_node(self, use_frame_pt: bool = False) -> CondNode:
+        walk_in_cond = Txn.application_args[0] == MethodSignature(self.method_sig)
+
+        if not (isinstance(self.condition, Expr) or self.condition == 1):
+            raise TealInputError("Invalid condition input for CondWithMethod")
+
+        res = ASTBuilder.wrap_handler(True, self.method, use_frame_pt=use_frame_pt)
+        if isinstance(self.condition, Expr):
+            res = Seq(Assert(self.condition), res)
+        return CondNode(walk_in_cond, res)
+
+
+CondWithMethod.__module__ = "pyteal"
+
+
 class ASTBuilder:
-    conditions_n_branches: list[CondNode] = field(default_factory=list)
+    def __init__(self):
+        self.methods_with_conds: list[CondWithMethod] = []
+        self.bare_calls: list[CondNode] = []
+
+    def _clean_bare_calls(self) -> None:
+        self.bare_calls = []
+
+    @staticmethod
+    def __filter_invalid_handlers_and_typecast(
+        subroutine: ABIReturnSubroutine | SubroutineFnWrapper | Expr,
+    ) -> ABIReturnSubroutine:
+        """This method filters out invalid handlers that might be normal subroutine, Expr, or unroutable ABIReturnSubroutine.
+
+        It accepts only routable ABIReturnSubroutine, and shrink the type to ABIReturnSubroutine from argument's union type.
+        """
+        if not isinstance(subroutine, ABIReturnSubroutine):
+            raise TealInputError(
+                f"method call should be only registering ABIReturnSubroutine, got {type(subroutine)}."
+            )
+        if not subroutine.is_abi_routable():
+            raise TealInputError(
+                f"method call ABIReturnSubroutine is not routable: "
+                f"got {subroutine.subroutine.argument_count()} args "
+                f"with {len(subroutine.subroutine.abi_args)} ABI args."
+            )
+        return subroutine
+
+    @staticmethod
+    def __subroutine_argument_instance_generate(
+        subroutine: ABIReturnSubroutine,
+    ) -> tuple[list[abi.BaseType], list[abi.BaseType], list[abi.Transaction]]:
+        # All subroutine args types
+        type_specs = cast(list[abi.TypeSpec], subroutine.subroutine.expected_arg_types)
+
+        # All subroutine arg values, initialize here and use below instead of
+        # creating new instances on the fly, so we don't have to think about splicing
+        # back in the transaction types
+        arg_vals = [typespec.new_instance() for typespec in type_specs]
+
+        # Only args that appear in app args
+        app_arg_vals: list[abi.BaseType] = [
+            ats for ats in arg_vals if not isinstance(ats, abi.Transaction)
+        ]
+
+        # only transaction args (these are omitted from app args)
+        txn_arg_vals: list[abi.Transaction] = [
+            ats for ats in arg_vals if isinstance(ats, abi.Transaction)
+        ]
+
```

Comment:
current transaction types:
    TransactionTypeSpec(),
    PaymentTransactionTypeSpec(),
    KeyRegisterTransactionTypeSpec(),
    AssetConfigTransactionTypeSpec(),
    AssetFreezeTransactionTypeSpec(),
    AssetTransferTransactionTypeSpec(),
    ApplicationCallTransactionTypeSpec(),

Reply:
right, I think `TransactionTypeSpec` is rather generic, while the others are specific (concrete).